File: Objective-C/ERStopWatch.m
//
//  ERStopWatch.m
//
//  Created by Zhang Zhicheng on 12-8-27.

#import "ERStopWatch.h"
#import <mach/mach_time.h>

@interface ERStopWatchModel : NSObject
@property (nonatomic, assign) UInt64 startTick;
@property (nonatomic, assign) double_t seconds;
@property (nonatomic, assign) UInt64 offset;
@property (nonatomic, assign) ERStopWatchState state;
@end

@implementation ERStopWatchModel
@end

@implementation ERStopWatch

static NSMutableDictionary<NSString *, ERStopWatchModel *> *watches;
static NSLock *watchLock;

+ (void)initialize
{
    if (self != ERStopWatch.class) {
        return;
    }

    watches = [NSMutableDictionary dictionary];
    watchLock = [[NSLock alloc] init];
}

+ (void)startWatch:(NSString *)watchName blk:(ERStopWatchBlk)blk
{
    [self withLock:^{
        ERStopWatchModel *watch = [[ERStopWatchModel alloc] init];
        watch.startTick = [self tick];
        watch.offset = 0;
        watch.state = ERStopWatchStateStart;
        watches[watchName] = watch;

        NSLog(@"------------- %@ : start", watchName);
        if (blk) blk(ERStopWatchStateStart, watchName, 0);
    }];
}

+ (void)stopWatch:(NSString *)watchName blk:(ERStopWatchBlk)blk
{
    [self withLock:^{
        ERStopWatchModel *watch = watches[watchName];
        if (!watch) return;

        NSArray<NSNumber *> *timing = [self secondsAndTicksForWatch:watch];
        double seconds = timing.firstObject.doubleValue;

        watch.state = ERStopWatchStateStop;
        watch.seconds = seconds;

        NSLog(@"------------- %@ : stop, total time %lf", watchName, seconds);
        if (blk) blk(ERStopWatchStateStop, watchName, seconds);
    }];
}

+ (void)cutWatch:(NSString *)watchName blk:(ERStopWatchBlk)blk
{
    [self cutWatch:watchName tag:nil blk:blk];
}

+ (void)cutWatch:(NSString *)watchName tag:(NSString *)tag blk:(ERStopWatchBlk)blk
{
    [self withLock:^{
        ERStopWatchModel *watch = watches[watchName];
        if (!watch) return;

        double seconds = [self secondsAndTicksForWatch:watch].firstObject.doubleValue;

        NSLog(@"------------- %@ : cut, tag %@ , time from start %lf", watchName, tag ?: @"", seconds);
        if (blk) blk(watch.state, watchName, seconds);
    }];
}

+ (void)pauseWatch:(NSString *)watchName blk:(ERStopWatchBlk)blk
{
    [self withLock:^{
        ERStopWatchModel *watch = watches[watchName];
        if (!watch || watch.state != ERStopWatchStateStart) return;

        NSArray<NSNumber *> *timing = [self secondsAndTicksForWatch:watch];
        double seconds = timing.firstObject.doubleValue;
        watch.offset = timing.lastObject.unsignedLongLongValue;
        watch.state = ERStopWatchStatePause;

        NSLog(@"------------- %@ : pause , time from start %lf", watchName, seconds);
        if (blk) blk(ERStopWatchStatePause, watchName, seconds);
    }];
}

+ (void)resumeWatch:(NSString *)watchName blk:(ERStopWatchBlk)blk
{
    [self withLock:^{
        ERStopWatchModel *watch = watches[watchName];
        if (!watch || watch.state != ERStopWatchStatePause) return;

        watch.startTick = [self tick];
        watch.state = ERStopWatchStateStart;

        NSLog(@"------------- %@ : resume", watchName);
        if (blk) blk(ERStopWatchStateStart, watchName, 0);
    }];
}

+ (void)withLock:(dispatch_block_t)work
{
    [watchLock lock];
    @try {
        work();
    } @finally {
        [watchLock unlock];
    }
}

+ (UInt64)tick
{
    return mach_absolute_time();
}

+ (NSArray<NSNumber *> *)secondsAndTicksForWatch:(ERStopWatchModel *)watch
{
    static mach_timebase_info_data_t timebase;
    if (timebase.denom == 0) {
        mach_timebase_info(&timebase);
    }

    UInt64 ticks = watch.offset + [self tick] - watch.startTick;
    double seconds = (double)ticks * 1e-9 * (double)timebase.numer / (double)timebase.denom;
    return @[@(seconds), @(ticks)];
}

@end

File: Sources/ERStopWatch/ERStopWatch.swift
import Foundation

public enum ERStopWatchState {
    case start, pause, stop
}

public typealias ERStopWatchBlk = (ERStopWatchState, String, Double) -> Void

public struct ERStopWatchSwift {
    static public func start(watchName: String, blk: ERStopWatchBlk?) {
        withLock {
            watches[watchName] = Watch(startTick: tick(), offset: 0, state: .start)
            print("------------- \(watchName) : start")
            blk?(.start, watchName, 0)
        }
    }

    static public func stop(watchName: String, blk: ERStopWatchBlk?) {
        withLock {
            guard var watch = watches[watchName] else { return }
            let (seconds, _) = elapsed(for: watch)

            watch.state = .stop
            watch.seconds = seconds
            watches[watchName] = watch

            print("------------- \(watchName) : stop , total time \(seconds)")
            blk?(.stop, watchName, seconds)
        }
    }

    static public func cut(watchName: String, tag: String = "", blk: ERStopWatchBlk?) {
        withLock {
            guard let watch = watches[watchName] else { return }
            let (seconds, _) = elapsed(for: watch)

            print("------------- \(watchName) : cut \(tag.isEmpty ? "" : "[\(tag)]") , time from start \(seconds)")
            blk?(watch.state, watchName, seconds)
        }
    }

    static public func pause(watchName: String, blk: ERStopWatchBlk?) {
        withLock {
            guard var watch = watches[watchName], watch.state == .start else { return }
            let (seconds, ticks) = elapsed(for: watch)

            watch.offset = ticks
            watch.state = .pause
            watches[watchName] = watch

            print("------------- \(watchName) : pause , time from start \(seconds)")
            blk?(.pause, watchName, seconds)
        }
    }

    static public func resume(watchName: String, blk: ERStopWatchBlk?) {
        withLock {
            guard var watch = watches[watchName], watch.state == .pause else { return }

            watch.startTick = tick()
            watch.state = .start
            watches[watchName] = watch

            print("------------- \(watchName) : resume")
            blk?(.start, watchName, 0)
        }
    }

    static private var watches = [String: Watch]()
    static private let lock = NSLock()

    static private func withLock(_ work: () -> Void) {
        lock.lock()
        defer { lock.unlock() }
        work()
    }

    static private func tick() -> UInt64 {
        DispatchTime.now().uptimeNanoseconds
    }

    static private func elapsed(for watch: Watch) -> (Double, UInt64) {
        let ticks = watch.offset + tick() - watch.startTick
        return (Double(ticks) * 1e-9, ticks)
    }
}

private struct Watch {
    var startTick: UInt64 = 0
    var seconds: Double = 0
    var offset: UInt64 = 0
    var state: ERStopWatchState
}
